src/persist/disk.rs

```diff
@@ -0,0 +1,118 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use std::fs::{self, File, OpenOptions};
+use std::io::{self, Read, Seek, SeekFrom, Write};
+use std::path::PathBuf;
+
+use appendix::Index;
+use canonical::{Canon, CanonError, Id, Source};
+
+use crate::generic::GenericTree;
+use crate::persist::{Backend, PersistError, PutResult};
+
+/// A disk-store for persisting microkelvin compound structures
+pub struct DiskBackend {
+    index: Index<Id, u64>,
+    data_path: PathBuf,
+    data_ofs: u64,
+}
+
+impl std::fmt::Debug for DiskBackend {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "DiskBackend {{ path: {:?}, data_ofs: {:?} }}",
+            self.data_path, self.data_ofs
+        )
+    }
+}
+
+impl DiskBackend {
+    /// Create a new disk backend
+    pub fn new<P>(path: P) -> io::Result<Self>
+    where
+        P: Into<PathBuf>,
+    {
+        let path = path.into();
+
+        let mut index_path = path.clone();
+        let mut data_path = path.clone();
+
+        index_path.push("index");
+        data_path.push("data");
+
+        fs::create_dir(&index_path)?;
+
+        let index = Index::new(&index_path)?;
+
+        let mut data = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(&data_path)?;
+
+        data.seek(SeekFrom::End(0))?;
+
+        let data_ofs = data.metadata()?.len();
+
+        Ok(DiskBackend {
+            data_path,
+            index,
+            data_ofs,
+        })
+    }
+}
+
+impl Backend for DiskBackend {
+    fn get(&self, id: &Id) -> Result<GenericTree, PersistError> {
+        if let Some(ofs) = self.index.get(id)? {
+            let mut data = File::open(&self.data_path)?;
+
+            data.seek(SeekFrom::Start(*ofs))?;
+
+            let len = id.size();
+
+            let mut buf = Vec::with_capacity(len);
+            buf.resize_with(len, || 0);
```

Comment:
You can just use `let mut buf = vec![0; len];`

Reply:
true.